File: apps/api/app/schemas/appsumo.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class RedeemRequest(BaseModel):
    code: str = Field(min_length=3, max_length=64)


class AppSumoCodePublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    code: str
    redeemed_at: datetime | None = None


class AppSumoStatus(BaseModel):
    """Current AppSumo lifetime entitlement for a workspace."""

    tier: int
    codes_redeemed: int
    max_tier: int
    can_stack_more: bool
    plan_code: str | None = None
    plan_display_name: str | None = None
    codes: list[AppSumoCodePublic] = []


# --- Admin (superuser) ------------------------------------------------------


class AdminGenerateRequest(BaseModel):
    count: int = Field(ge=1, le=10_000)
    batch: str | None = Field(default=None, max_length=64)
    prefix: str = Field(default="ADV", max_length=8)


class AdminGenerateResponse(BaseModel):
    generated: int
    batch: str | None = None
    codes: list[str]


class CodeStats(BaseModel):
    total: int
    redeemed: int
    refunded: int
    unredeemed: int


class DeactivateRequest(BaseModel):
    code: str = Field(min_length=3, max_length=64)

File: apps/web/src/lib/appsumo.ts
import { apiFetch } from "@/lib/api-client";

export type AppSumoCode = {
  code: string;
  redeemed_at: string | null;
};

export type AppSumoStatus = {
  tier: number;
  codes_redeemed: number;
  max_tier: number;
  can_stack_more: boolean;
  plan_code: string | null;
  plan_display_name: string | null;
  codes: AppSumoCode[];
};

export function getAppSumoStatus(workspaceId: string) {
  return apiFetch<AppSumoStatus>(`/workspaces/${workspaceId}/appsumo/status`);
}

export function redeemAppSumoCode(workspaceId: string, code: string) {
  return apiFetch<AppSumoStatus>(`/workspaces/${workspaceId}/appsumo/redeem`, {
    method: "POST",
    body: { code },
  });
}
